test/e2e/balancerSwap.test.ts

```diff
@@ -0,0 +1,481 @@
+import ERC20 from "@openzeppelin/contracts/build/contracts/ERC20PresetMinterPauser.json";
+import { expect } from "chai";
+import Debug from "debug";
+import { BigNumberish, Contract, Wallet } from "ethers";
+import { ethers, waffle } from "hardhat";
+
+import {
+  OrderKind,
+  SwapEncoder,
+  SigningScheme,
+  TypedDataDomain,
+  domain,
+} from "../../src/ts";
+
+import MockPool from "./balancer/MockPool.json";
+import { deployTestContracts } from "./fixture";
+
+const LOTS = ethers.utils.parseEther("10000.0");
+const debug = Debug("e2e:balancerSwap");
+
+describe("E2E: Direct Balancer swap", () => {
+  let deployer: Wallet;
+  let solver: Wallet;
+  let pooler: Wallet;
+  let trader: Wallet;
+
+  let vault: Contract;
+  let settlement: Contract;
+  let vaultRelayer: Contract;
+  let domainSeparator: TypedDataDomain;
+
+  let tokens: [Contract, Contract, Contract];
+  let pools: Record<string, Contract>;
+
+  let snapshot: unknown;
+
+  before(async () => {
+    const deployment = await deployTestContracts();
+
+    ({
+      deployer,
+      vault,
+      settlement,
+      vaultRelayer,
+      wallets: [solver, pooler, trader],
+    } = deployment);
+
+    const { vaultAuthorizer, authenticator, manager } = deployment;
+    for (const method of [
+      "batchSwapGivenIn",
+      "batchSwapGivenOut",
+      "transferInternalBalance",
+      "transferToExternalBalance",
+      "withdrawFromInternalBalance",
+    ]) {
+      await vaultAuthorizer
+        .connect(manager)
+        .grantRole(
```

Comment:
Is there a way to grant multiple roles in one transaction? Otherwise it might be pretty inefficient for EOAs

Reply:
This is done by the `manager` i.e. the vault admin, and is only done once. The EOA does a single `changeRelayerApproval` call which grants all permissions given by the `grantRole` calls.